File: framework/languages/dotnet/e2e/PubSub/Client/Scenarios/PsC1MissingMessageNameScenario.cs
// Verifies PS-C1 Missing Message Name behavior.
using PubSub.Shared;
using Zlink.HttpClient;

namespace PubSub.Client.Scenarios;

// PS-C1: verifies that an unregistered publish message is dropped and normal delivery continues.
internal static class PsC1MissingMessageNameScenario
{
    public static async Task RunAsync(ZLinkHttpClient publisher, IReadOnlyList<ZLinkHttpClient> subscribers)
    {
        var runId = Guid.NewGuid().ToString("N");

        // Publish a packet name that subscribers did not register, so dispatch should drop it.
        await publisher.Post("/publish/missing")
            .Query("topic", PubSubNames.MainTopic)
            .Query("runId", runId)
            .Query("sequence", "1")
            .Query("value", "missing")
            .AsyncRaw();

        // The drop is observed on subscriber evidence because publisher submit only reaches transport.
        await WaitForSubscribersAsync(subscribers, new EvidenceWaitReq(
            [],
            [])
        {
            ContainsAllLineGroups =
            [
                ["dispatch-error|", "packet=MissingEventMsg", $"topic={PubSubNames.MainTopic}"]
            ]
        });

        // A normal event after the missing handler case proves the channel continues to work.
        await publisher.Post("/publish/event")
            .Query("topic", PubSubNames.MainTopic)
            .Query("runId", runId)
            .Query("sequence", "2")
            .Query("value", "after-missing")
            .AsyncRaw();
        await WaitForSubscribersAsync(subscribers, new EvidenceWaitReq(
            ["event|", $"run={runId}", $"topic={PubSubNames.MainTopic}"],
            []));

        Console.WriteLine("scenario PS-C1 passed");
    }

    private static async Task WaitForSubscribersAsync(
        IReadOnlyList<ZLinkHttpClient> subscribers,
        EvidenceWaitReq request)
    {
        var waits = subscribers
            .Select(subscriber => subscriber.Post("/evidence/wait").Body(request).Async<string[]>().AsTask())
            .ToArray();
        await Task.WhenAll(waits);
    }
}

File: framework/languages/node/e2e/PubSub/Client/Scenarios/ps-c1-missing-message-name-scenario.ts
// PS-C1: 등록되지 않은 packet name을 drop한다 시나리오를 검증한다.
import { randomUUID } from 'node:crypto';
import { PacketNames, PubSubNames } from '../../Shared/messages';
import { postJson } from '../../../http-client';
import { publishEvent, waitForAll } from './ps-a1-fanout-basic-delivery-scenario';

export async function runPsC1(publisher: string, subscribers: readonly string[]): Promise<void> {
  const runId = randomUUID().replaceAll('-', '');
  await postJson(publisher, '/publish/missing', {
    topic: PubSubNames.mainTopic,
    runId,
    sequence: 1,
    value: 'missing'
  });

  await waitForAll(subscribers, {
    containsAllLineGroups: [['dispatch-error|', `packet=${PacketNames.missingEventMsg}`, `topic=${PubSubNames.mainTopic}`]],
    timeoutMilliseconds: 10_000
  });

  await publishEvent(publisher, PubSubNames.mainTopic, runId, 2, 'after-missing');
  await waitForAll(subscribers, {
    containsAll: ['event|', `run=${runId}`, `topic=${PubSubNames.mainTopic}`],
    timeoutMilliseconds: 10_000
  });
  console.log('scenario PS-C1 passed');
}
